File: Design Patterns/Behavioral/Command.ts
interface ICommand {
    execute(): void
    undo(): void
}

class Light {
    turnOn() { console.log("Light ON") }
    turnOff() { console.log("Light OFF") }
}
class LightOnCommand implements ICommand {
    constructor(private light: Light) { }

    execute() { this.light.turnOn() }
    undo() { this.light.turnOff() }
}

class LightOffCommand implements ICommand {
    constructor(private light: Light) { }

    execute() { this.light.turnOff() }
    undo() { this.light.turnOn() }
}

class RemoteControl {
    private history: ICommand[] = []

    pressButton(command: ICommand) {
        command.execute()
        this.history.push(command)
    }

    undo() {
        const command = this.history.pop()
        if (command) { command.undo() }
    }
}

const light = new Light()

const lightOn = new LightOnCommand(light)
const lightOff = new LightOffCommand(light)

const remote = new RemoteControl()

remote.pressButton(lightOn)
remote.pressButton(lightOff)

remote.undo()
remote.undo() 

File: Design Patterns/Behavioral/C#/Command.cs
using System;
using System.Collections.Generic;
interface ICommand
{
  void Execute();
  void Undo();
}

class Light
{
  public void TurnOn() { Console.WriteLine("Light ON"); }
  public void TurnOff() { Console.WriteLine("Light OFF"); }
}
class LightOnCommand : ICommand
{
  private readonly Light light;

  public LightOnCommand(Light light)
  {
    this.light = light;
  }


  public void Execute() { this.light.TurnOn(); }
  public void Undo() { this.light.TurnOff(); }
}

class LightOffCommand : ICommand
{
  private readonly Light light;

  public LightOffCommand(Light light)
  {
    this.light = light;
  }

  public void Execute() { this.light.TurnOff(); }
  public void Undo() { this.light.TurnOn(); }
}

class RemoteControl
{
  private List<ICommand> history = new List<ICommand>();

  public void PressButton(ICommand command)
  {
    command.Execute();
    this.history.Add(command);
  }

  public void Undo()
  {
    if (this.history.Count == 0) return;

    var command = this.history[this.history.Count - 1];
    this.history.RemoveAt(this.history.Count - 1);

    if (command != null) { command.Undo(); }
  }
}
class Program
{
  public static void Main(string[] args)
  {

    Light light = new Light();

    LightOnCommand lightOn = new LightOnCommand(light);
    LightOffCommand lightOff = new LightOffCommand(light);

    RemoteControl remote = new RemoteControl();

    remote.PressButton(lightOn);
    remote.PressButton(lightOff);

    remote.Undo();
    remote.Undo();
  }
}